miden-lib/src/notes/mod.rs

```diff
@@ -30,12 +31,12 @@ pub enum Script {
 /// 1. P2ID - pay to id.
 /// 2. P2IDR - pay to id with recall after a certain block height.
 /// 3. SWAP - swap of assets between two accounts.
-pub fn create_note(
+pub fn create_note<R: FeltRng>(
     script: Script,
     assets: Vec<Asset>,
     sender: AccountId,
     tag: Option<Felt>,
-    serial_num: Word,
+    mut rng: R,
 ) -> Result<Note, NoteError> {
     let note_assembler = assembler();
```

Comment:
With this, we can also avoid including `serial_num` in `SWAP` script (line 26) above.

Reply:
My understanding is that the serial_num used is the SWAP script needs to be inputed by the sender manually so he can know the pre-image to the RECIPIENT and consume the tx.

If we make this number random it might mean that the sender of the initial `SWAP` script will not be able to recover the funds of the generated P2ID. 

Am I mistaken? What are you thoughts?

Reply:
Ah - good point! I wonder if there is a good way to change the API of this function so that we can still generate the `serial_num` internally, but also return the returning note info to the caller. (will think more about it).

Reply:
For example, if this was a specific function for creating atomic swaps, it could have been:
```Rust
pub fn create_swap<R: FeltRng>(
    out_asset: Asset,
    in_asset: Asset,
    sender: AccountId,
    mut rng: R,
) -> Result<(Note, Note), NoteError>
```
Where the returned tuple contains both outgoing and returning notes.

We handle the input parameters into this function generically via the `Script` enum. But not sure how to handle output parameters generically.

Reply:
Would it also mean  that we wouldn't have like now a general-purpose function using a match for different scripts but potentially have a whole function implemented for each different script managing the specific behaviour of this script?

Reply:
Here I don't see reference of a Recipient. Hence can the initial `SWAP` sender be able to recover the funds just by having knowledge of the 2 produced notes?

Reply:
> Here I don't see reference of a Recipient. Hence can the initial `SWAP` sender be able to recover the funds just by having knowledge of the 2 produced notes?

Yes, we can get `recipient` from the `Note` object.

Reply:
Is the `Note` object the one we send on the network? 

If yes then it would mean that anyone having access to the note can consume it.

Reply:
> Is the `Note` object the one we send on the network?

It depends, for public transactions - yes; for private transactions, no (we just send a commitment to the note).

> If yes then it would mean that anyone having access to the note can consume it.

This is actually not a problem. For the outgoing note, we want anyone to be able to consume it (we don't care who does it as long as they are willing to swap the assets). For the returning note, the `recipient` specifies that only the account which created the swap can consume it.

Reply:
Understood.

Reply:
Thinking more about this, I wonder if we need to have the `Script` enum at all - or at least whether we need to use it for constructing `Note` objects. An alternative would be to use a separate function per script. For example:
```Rust
pub fn create_p2id_note<R: FeltRng>(
    sender: AccountId,
    target: AccountId,
    assets: Vec<Asset>,
    mut rng: R,
) -> Result<Note, NoteError>

pub fn create_p2idr_note<R: FeltRng>(
    sender: AccountId,
    target: AccountId,
    assets: Vec<Asset>,
    recall_height: u32,
    mut rng: R,
) -> Result<(Note, Note), NoteError>

pub fn create_swap_note<R: FeltRng>(
    sender: AccountId,
    offered_asset: Asset,
    requested_asset: Asset,
    mut rng: R,
) -> Result<(Note, Note), NoteError>
```

Reply:
How would we manage the `tag` in this situation?

After testing an implementation using separate functions per script and removing the script enum I believe that this update adds a quantity of duplicated code and makes the code less readable.

My understanding of our current implementation is the following: 

We have flexibility in the script inputs using the enum fields: https://github.com/0xPolygonMiden/miden-base/blob/84811c7d70cea8fb88af8901f4a9452c58044665/miden-lib/src/notes/mod.rs#L19-L26

We prevent code duplication using a general purpose function: 
(When I say "once" I mean that we don't duplicate code)
- Flexibility over the `tag`
- Clarity over the different possible note_scripts
- Note generation exists in one location
- Declaring the `assembler` once
- Declaring the `note_script` once
- Creating the note once

https://github.com/0xPolygonMiden/miden-base/blob/84811c7d70cea8fb88af8901f4a9452c58044665/miden-lib/src/notes/mod.rs#L28-L81

But with this method we don't have much flexibility in the return of the function considering it should be the same for all scripts (I have understood that you found that a problem of the current impl): 
```rust
Result<Note, NoteError>
```

This method might also not be very scalable in the future in case we add a large quantity of note scripts.

I might not fully understand the reason why we would want create separate functions for script creation as of now, could you please explain?